File: cli/src/vx-parser/read.ts
import { writeFileSync, readFileSync, readdirSync } from "fs";
import { OrderedSet } from "immutable";
import { load } from "@hyrious/marshal";
import { inflateSync } from "zlib";

import { getValueBySymbolDesc } from "./symbol-utils";

export function readMap(inputDir: string, outputDir: string, logging: boolean, logString: string): void {
    const files = readdirSync(inputDir).filter(
        (filename) => filename.startsWith("Map") && !filename.startsWith("MapInfos")
    );

    const objMap = new Map(
        files.map((filename) => [filename, load(readFileSync(`${inputDir}/${filename}`)) as object])
    );

    const lines = OrderedSet().asMutable() as OrderedSet<string>;
    const namesLines = OrderedSet().asMutable() as OrderedSet<string>;

    for (const [filename, obj] of objMap) {
        const displayName = getValueBySymbolDesc(obj, "@display_name");

        if (displayName) {
            namesLines.add(displayName);
        }

        const events: object = getValueBySymbolDesc(obj, "@events");

        for (const event of Object.values(events || {})) {
            const pages: object[] = getValueBySymbolDesc(event, "@pages");
            if (!pages) {
                continue;
            }

            for (const page of pages) {
                let inSeq: boolean = false;
                const line: string[] = [];
                const list: object[] = getValueBySymbolDesc(page, "@list");

                for (const item of list) {
                    const code: number = getValueBySymbolDesc(item, "@code");
                    const parameters: string[] = getValueBySymbolDesc(item, "@parameters");

                    for (const parameter of parameters) {
                        if (code === 401) {
                            inSeq = true;

                            if (typeof parameter === "string") {
                                line.push(parameter);
                            }
                        } else {
                            if (inSeq) {
                                const lineJoined = line.join("\\n");

                                lines.add(lineJoined);
                                line.length = 0;
                                inSeq = false;
                            }

                            switch (code) {
                                case 102:
                                    if (Array.isArray(parameter)) {
                                        for (const param of parameter) {
                                            if (typeof param === "string") {
                                                lines.add(param);
                                            }
                                        }
                                    }
                                    break;

                                case 356:
                                    if (
                                        typeof parameter === "string" &&
                                        parameter.startsWith("GabText") &&
                                        parameter.startsWith("choice_text") &&
                                        !parameter.endsWith("????")
                                    ) {
                                        lines.add(parameter);
                                    }
                                    break;
                            }
                        }
                    }
                }
            }
        }

        if (logging) {
            console.log(`${logString} ${filename}`);
        }
    }

    writeFileSync(`${outputDir}/maps.txt`, lines.join("\n"), "utf8");
    writeFileSync(`${outputDir}/maps_trans.txt`, "\n".repeat(lines.size ? lines.size - 1 : 0), "utf8");
    writeFileSync(`${outputDir}/names.txt`, namesLines.join("\n"), "utf8");
    writeFileSync(`${outputDir}/names_trans.txt`, "\n".repeat(namesLines.size ? namesLines.size - 1 : 0), "utf8");
}

export function readOther(inputDir: string, outputDir: string, logging: boolean, logString: string): void {
    const filenames = readdirSync(inputDir).filter((filename: string) => {
        const files = ["Map", "Tilesets", "Animations", "States", "System", "Plugins", "Scripts"];
        return files.some((file) => filename.startsWith(file)) ? false : true;
    });

    const objArrMap = new Map(
        filenames.map((filename) => [filename, load(readFileSync(`${inputDir}/${filename}`)) as object[]])
    );

    for (const [filename, objArr] of objArrMap) {
        const lines = OrderedSet().asMutable() as OrderedSet<string>;

        if (!filename.startsWith("Common") && !filename.startsWith("Troops")) {
            for (const obj of objArr.slice(1)) {
                const name: string = getValueBySymbolDesc(obj, "@name");
                const nickname: string = getValueBySymbolDesc(obj, "@nickname");
                const description: string = getValueBySymbolDesc(obj, "@description");
                const note: string = getValueBySymbolDesc(obj, "@note");

                if (typeof name === "string" && name) {
                    lines.add(name);
                }

                if (typeof nickname === "string" && nickname) {
                    lines.add(nickname);
                }

                if (typeof description === "string" && description) {
                    lines.add(description.replaceAll(/\r\n|\n(?!bt)/g, "\\n"));
                }

                if (typeof note === "string" && note) {
                    lines.add(note.replaceAll(/\r\n|\n(?!bt)/g, "\\n"));
                }
            }

            writeFileSync(
                `${outputDir}/${filename.toLowerCase().slice(0, filename.lastIndexOf("."))}.txt`,
                lines.join("\n"),
                "utf8"
            );
            writeFileSync(
                `${outputDir}/${filename.toLowerCase().slice(0, filename.lastIndexOf("."))}_trans.txt`,
                "\n".repeat(lines.size ? lines.size - 1 : 0),
                "utf8"
            );
            continue;
        } else {
            for (const obj of objArr.slice(1)) {
                const pages: object[] = getValueBySymbolDesc(obj, "@pages");
                const pagesLength = pages ? pages.length : 1;

                for (let i = 0; i < pagesLength; i++) {
                    const list: object[] =
                        pagesLength > 1 ? getValueBySymbolDesc(pages[i], "@list") : getValueBySymbolDesc(obj, "@list");

                    if (!Array.isArray(list)) {
                        continue;
                    }

                    let inSeq: boolean = false;
                    const line: string[] = [];

                    for (const item of list) {
                        const code: number = getValueBySymbolDesc(item, "@code");
                        const parameters: string[] = getValueBySymbolDesc(item, "@parameters");

                        for (const parameter of parameters) {
                            if (code === 401 || code === 405) {
                                inSeq = true;

                                if (typeof parameter === "string") {
                                    line.push(parameter);
                                }
                            } else {
                                if (inSeq) {
                                    const lineJoined = line.join("\\n");

                                    lines.add(lineJoined);
                                    line.length = 0;
                                    inSeq = false;
                                }

                                switch (code) {
                                    case 102:
                                        if (Array.isArray(parameter)) {
                                            for (const param of parameter) {
                                                if (typeof param === "string") {
                                                    lines.add(param);
                                                }
                                            }
                                        }
                                        break;

                                    case 356:
                                        if (
                                            typeof parameter === "string" &&
                                            parameter.startsWith("GabText") &&
                                            parameter.startsWith("choice_text") &&
                                            !parameter.endsWith("????")
                                        ) {
                                            lines.add(parameter);
                                        }
                                        break;
                                }
                            }
                        }
                    }
                }
            }
        }

        if (logging) {
            console.log(`${logString} ${filename}`);
        }

        writeFileSync(
            `${outputDir}/${filename.toLowerCase().slice(0, filename.lastIndexOf("."))}.txt`,
            lines.join("\n"),
            "utf8"
        );

        writeFileSync(
            `${outputDir}/${filename.toLowerCase().slice(0, filename.lastIndexOf("."))}_trans.txt`,
            "\n".repeat(lines.size ? lines.size - 1 : 0),
            "utf8"
        );
    }
}

export function readSystem(inputDir: string, outputDir: string, logging: boolean, logString: string): void {
    const systemPath = `${inputDir}/System.rvdata2`;
    const obj = load(readFileSync(systemPath)) as object;

    const lines = OrderedSet().asMutable() as OrderedSet<string>;
    const symbols = ["@skill_types", "@weapon_types", "@armor_types", "@currency_unit", "@terms"];

    const [skillTypes, weaponTypes, armorTypes, currencyUnit, terms] = symbols.map((symbol) =>
        getValueBySymbolDesc(obj, symbol)
    );

    for (const arr of [skillTypes, weaponTypes, armorTypes]) {
        for (const string of arr) {
            if (string) {
                lines.add(string);
            }
        }
    }

    lines.add(currencyUnit);

    const termsSymbols = Object.getOwnPropertySymbols(terms);

    for (let i = 0; i < termsSymbols.length; i++) {
        for (const string of terms[termsSymbols[i]] as string[]) {
            if (string) {
                lines.add(string);
            }
        }
    }

    if (logging) {
        console.log(`${logString} system.txt`);
    }

    writeFileSync(`${outputDir}/system.txt`, lines.join("\n"), "utf8");
    writeFileSync(`${outputDir}/system_trans.txt`, "\n".repeat(lines.size ? lines.size - 1 : 0), "utf8");
}

export function readScripts(inputDir: string, outputDir: string, logging: boolean, logString: string): void {
    const scriptsPath = `${inputDir}/Scripts.rvdata2`;
    const uintarrArr = load(readFileSync(scriptsPath), { string: "binary" }) as Uint8Array[][];

    const fullCode = [];
    for (const [_m, _t, code] of uintarrArr) {
        const codeString = inflateSync(code)
            .toString("utf8")
            .replaceAll(/\r\n|\n(?!bt)/g, "\\n");
        fullCode.push(codeString);
    }

    if (logging) {
        console.log(`${logString} scripts.txt`);
    }

    const joinedCode = fullCode.join("\n");
    writeFileSync(`${outputDir}/scripts.txt`, joinedCode, "utf8");
    writeFileSync(`${outputDir}/scripts_trans.txt`, joinedCode, "utf8");
}

File: gui/src-tauri/src/read.rs
use fancy_regex::Regex;
use fnv::{FnvBuildHasher, FnvHashMap};
use indexmap::IndexSet;
use rayon::prelude::*;
use serde_json::{from_str, Value};
use std::{
    fs::{read_dir, read_to_string, write, DirEntry},
    path::{Path, PathBuf},
};

pub fn read_map(original_path: &Path, output_path: &Path) {
    let re: Regex = Regex::new(r"^Map[0-9].*.json$").unwrap();

    let files: Vec<DirEntry> = read_dir(original_path)
        .unwrap()
        .flatten()
        .filter(|entry: &DirEntry| {
            let filename: String = entry.file_name().into_string().unwrap();
            re.is_match(&filename).unwrap()
        })
        .collect();

    let maps_obj_map: Vec<Value> = files
        .iter()
        .map(|entry: &DirEntry| from_str(&read_to_string(entry.path()).unwrap()).unwrap())
        .collect();

    let mut lines: IndexSet<String, FnvBuildHasher> = IndexSet::default();
    let mut names_lines: IndexSet<String, FnvBuildHasher> = IndexSet::default();

    for obj in maps_obj_map {
        if let Some(display_name) = obj["displayName"].as_str() {
            if !display_name.is_empty() {
                names_lines.insert(display_name.to_string());
            }
        }

        for event in obj["events"].as_array().unwrap().iter().skip(1) {
            if !event["pages"].is_array() {
                continue;
            }

            for page in event["pages"].as_array().unwrap().iter() {
                let mut in_seq: bool = false;
                let mut line: Vec<String> = Vec::new();

                for list in page["list"].as_array().unwrap() {
                    let code: u16 = list["code"].as_u64().unwrap() as u16;

                    for parameter_value in list["parameters"].as_array().unwrap() {
                        if code == 401 {
                            in_seq = true;

                            if parameter_value.is_string() {
                                let parameter: &str = parameter_value.as_str().unwrap();
                                line.push(parameter.to_string());
                            }
                        } else {
                            if in_seq {
                                lines.insert(line.join(r"\#"));
                                line.clear();
                                in_seq = false;
                            }

                            match code {
                                102 => {
                                    if parameter_value.is_array() {
                                        for param_value in parameter_value.as_array().unwrap() {
                                            if param_value.is_string() {
                                                let param: &str = param_value.as_str().unwrap();
                                                lines.insert(param.to_string());
                                            }
                                        }
                                    }
                                }

                                356 => {
                                    if parameter_value.is_string() {
                                        let parameter: &str = parameter_value.as_str().unwrap();

                                        if parameter.starts_with("GabText")
                                            && (parameter.starts_with("choice_text")
                                                && !parameter.ends_with("????"))
                                        {
                                            lines.insert(parameter.to_string());
                                        }
                                    }
                                }

                                _ => {}
                            }
                        }
                    }
                }
            }
        }
    }

    write(
        output_path.join("maps.txt"),
        lines.iter().cloned().collect::<Vec<String>>().join("\n"),
    )
    .unwrap();

    write(
        output_path.join("names.txt"),
        names_lines
            .iter()
            .cloned()
            .collect::<Vec<String>>()
            .join("\n"),
    )
    .unwrap();

    write(
        output_path.join("maps_trans.txt"),
        "\n".repeat(lines.len() - 1),
    )
    .unwrap();

    write(
        output_path.join("names_trans.txt"),
        "\n".repeat(names_lines.len() - 1),
    )
    .unwrap();
}

pub fn read_other(original_path: &Path, output_path: &Path) {
    let re: Regex = Regex::new(r"^(?!Map|Tilesets|Animations|States|System).*json$").unwrap();

    let files: Vec<DirEntry> = read_dir(original_path)
        .unwrap()
        .flatten()
        .filter(|entry: &DirEntry| {
            re.is_match(&entry.file_name().into_string().unwrap())
                .unwrap()
        })
        .collect();

    let obj_arr_map: FnvHashMap<String, Vec<Value>> = files
        .par_iter()
        .map(|entry: &DirEntry| {
            (
                entry.file_name().into_string().unwrap(),
                from_str(&read_to_string(entry.path()).unwrap()).unwrap(),
            )
        })
        .collect();

    for (filename, obj_arr) in obj_arr_map {
        let processed_filename: String = filename[0..filename.rfind('.').unwrap()].to_lowercase();
        let mut lines: IndexSet<String, FnvBuildHasher> = IndexSet::default();

        if !filename.starts_with("Common") && !filename.starts_with("Troops") {
            for obj in obj_arr {
                if obj["name"].is_string() {
                    let name: &str = obj["name"].as_str().unwrap();

                    if !name.is_empty() {
                        lines.insert(name.to_string());
                    }
                }

                if obj["description"].is_string() {
                    let description: &str = obj["description"].as_str().unwrap();

                    if !description.is_empty() {
                        lines.insert(description.replace('\n', r"\#"));
                    }
                }

                if obj["note"].is_string() {
                    let note: &str = obj["note"].as_str().unwrap();

                    if !note.is_empty() {
                        lines.insert(note.replace('\n', r"\#"));
                    }
                }
            }

            write(
                output_path.join(format!("{processed_filename}.txt",)),
                lines.iter().cloned().collect::<Vec<String>>().join("\n"),
            )
            .unwrap();
            write(
                output_path.join(format!("{processed_filename}_trans.txt",)),
                "\n".repeat(lines.len() - 1),
            )
            .unwrap();
            continue;
        }

        for obj in obj_arr.iter().skip(1) {
            let pages_length: usize = if filename.starts_with("Troops") {
                obj["pages"].as_array().unwrap().len()
            } else {
                1
            };

            for i in 0..pages_length {
                let list: &Value = if pages_length != 1 {
                    &obj["pages"][i]["list"]
                } else {
                    &obj["list"]
                };

                if !list.is_array() {
                    continue;
                }

                let mut in_seq: bool = false;
                let mut line: Vec<String> = Vec::new();

                for list in list.as_array().unwrap() {
                    let code: u16 = list["code"].as_u64().unwrap() as u16;

                    for parameter_value in list["parameters"].as_array().unwrap() {
                        if code == 401 || code == 405 {
                            in_seq = true;

                            if parameter_value.is_string() {
                                let parameter: &str = parameter_value.as_str().unwrap();
                                line.push(parameter.to_string());
                            }
                        } else {
                            if in_seq {
                                lines.insert(line.join(r"\#"));
                                line.clear();
                                in_seq = false;
                            }

                            match code {
                                102 => {
                                    if parameter_value.is_array() {
                                        for param_value in parameter_value.as_array().unwrap() {
                                            if param_value.is_string() {
                                                let param: &str = param_value.as_str().unwrap();
                                                lines.insert(param.to_string());
                                            }
                                        }
                                    }
                                }

                                356 => {
                                    if parameter_value.is_string() {
                                        let parameter: &str = parameter_value.as_str().unwrap();

                                        if parameter.starts_with("GabText")
                                            && (parameter.starts_with("choice_text")
                                                && !parameter.ends_with("????"))
                                        {
                                            lines.insert(parameter.to_string());
                                        }
                                    }
                                }

                                _ => {}
                            }
                        }
                    }
                }
            }
        }

        write(
            output_path.join(format!("{processed_filename}.txt")),
            lines.iter().cloned().collect::<Vec<String>>().join("\n"),
        )
        .unwrap();

        write(
            output_path.join(format!("{processed_filename}_trans.txt")),
            "\n".repeat(lines.len() - 1),
        )
        .unwrap();
    }
}

pub fn read_system(original_path: &Path, output_path: &Path) {
    let system_file: PathBuf = original_path.join("System.json");

    let obj: Value = from_str(&read_to_string(system_file).unwrap()).unwrap();

    let mut lines: IndexSet<String, FnvBuildHasher> = IndexSet::default();

    for string in obj["equipTypes"].as_array().unwrap() {
        if string.is_string() {
            let element_str: &str = string.as_str().unwrap();

            if !element_str.is_empty() {
                lines.insert(element_str.to_string());
            }
        }
    }

    for string in obj["skillTypes"].as_array().unwrap() {
        if string.is_string() {
            let element_str: &str = string.as_str().unwrap();

            if !element_str.is_empty() {
                lines.insert(string.as_str().unwrap().to_string());
            }
        }
    }

    for (key, value) in obj["terms"].as_object().unwrap() {
        if key != "messages" {
            for string in value.as_array().unwrap() {
                if string.is_string() {
                    let string_str: &str = string.as_str().unwrap();

                    if !string_str.is_empty() {
                        lines.insert(string_str.to_string());
                    }
                }
            }
        } else {
            if !value.is_object() {
                return;
            }

            for message_value in value.as_object().unwrap().values() {
                if message_value.is_string() {
                    let message_str: &str = message_value.as_str().unwrap();

                    if !message_str.is_empty() {
                        lines.insert(message_str.to_string());
                    }
                }
            }
        }
    }

    write(
        output_path.join("system.txt"),
        lines.iter().cloned().collect::<Vec<String>>().join("\n"),
    )
    .unwrap();
    write(
        output_path.join("system_trans.txt"),
        "\n".repeat(lines.len() - 1),
    )
    .unwrap();
}
